helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java

```diff
@@ -487,10 +490,18 @@ private void chargePendingTransition(Resource resource, CurrentStateOutput curre
           String pendingState = pendingMap.get(instance);
           if (pendingState != null && !pendingState.equals(currentState)
               && !cache.getDisabledInstancesForPartition(resourceName, partition.getPartitionName())
-                  .contains(instance)) {
+              .contains(instance)) {
             // Only charge this instance if the partition is not disabled
             throttleController.chargeInstance(rebalanceType, instance);
             shouldChargePartition = true;
+            // Remove the partition marked as partition need recovery/load balance. Otherwise, they
+            // will be double charged when they already charged for pending messages.
+            // At the same time, we need to update the intermediate map since it was supposed to
+            // be updated when the recovery rebalance throttling piece.
+            cleanPartitionsAreCharged(partition, partitionsNeedRecovery, partitionsNeedLoadbalance,
+                rebalanceType);
+            intermediatePartitionStateMap
```

Comment:
I think we could tweak the structure to minimize confusion in the future. Here's a suggestion:

- First, please use this for the inline comment:

// If there is a pending state transition for the partition, that means that an assignment has already been made and the state transition message has already been sent out for the partition in a previous pipeline run. We must honor this and reflect it by charging for the pending state transition message.
// Since the assignment has already been made for the pending message, we do a special treatment for it by setting the best possible state directly in intermediatePartitionStateMap so that the pending message won't be double-assigned or double-charged in recovery or load balance.

- And please merge the two lines of logic by creating a method (instead of cleanPartitionsAreCharged):

```
  /**
   * Pretreat a partition with a pending message so that the partition will not be double-charged or double-assigned during recovery and load balance.
   * @param partition
   * @param partitionsNeedRecovery
   * @param partitionsNeedLoadbalance
   * @param rebalanceType
   */
  private void preAssignPendingTransitions(Partition partition, Set<Partition> partitionsNeedRecovery,
      Set<Partition> partitionsNeedLoadbalance, RebalanceType rebalanceType, PartitionStateMap bestPossiblePartitionStateMap,
      PartitionStateMap intermediatePartitionStateMap) {
    // Pass the best possible state directly into intermediatePartitionStateMap
    // This is safe to do so because we already have a pending transition for this partition, implying that the assignment has been made in previous pipeline
    intermediatePartitionStateMap.setState(partition, bestPossiblePartitionStateMap.getPartitionMap(partition));
    // Remove the partition's name from the set of partition (names) that need to be charged and assigned to prevent double-processing
    switch (rebalanceType) {
    case RECOVERY_BALANCE:
      partitionsNeedRecovery.remove(partition);
      break;
    case LOAD_BALANCE:
      partitionsNeedLoadbalance.remove(partition);
      break;
    }
  }
```

I think this way we make it 100% clear what our intentions are and why we're making this change.

Reply:
Good comments and craftmanship, Hunter!

Reply:
Just make the name handlePendingStateTransitionsForThrottling to make it clear. Pretreat sounds more like a hacky way.

Reply:
@dasahcc Sure. I think that name makes sense.